docs/design/multipart-upload.md

```diff
@@ -0,0 +1,82 @@
+# Multipart Upload
+
+## Abstract
+
+This design document describes multipart upload and how it can be implemented.
+Currently, there is no direct support for multipart upload on the Satellite side.
+The feature is simulated on the gateway side, which requires keeping 
+a lot of data in memory and doesn't achieve the full usefulness of multipart upload.
+
+## Background
+
+Multipart upload gives quite a few nice properties:
+
+* Concurrent uploads,
+* continuation of partial uploads, and
+* out-of-order uploads.
+
+The flow for a multipart upload, from a S3 perspective, is:
+
+1. starts a new multipart upload,
+2. the upload data is split into multiple parts,
+3. for each part: send the part,
+4. calls "complete multipart upload," or
+5. alternatively, calls "cancel multipart upload."
+
+There are a few important things we must keep in mind:
+
+* the parts can have different sizes,
+* during the start of the upload, the part size can be unknown,
+* the parts can be uploaded from different computers,
+* the parts can be uploaded out of order, and
+* the parts can be reuploaded, in case of a failed upload.
+
+## Design
+
+We change the path handling to support listing parts uploaded in arbitrary order. When we start an upload we will create an object at the following path:
+
+```
+<project-id>/<bucket-id>/objects/<path ...>_<version> => object information (partial)
```

Comment:
Can we avoid adding new special characters (`_` in this case) to the path grammar?

It can be `<project-id>/<bucket-id>/objects/<path ...>/<version>`. The rule can be that the last path element is the version.

Reply:
I think that would make distinguishing folders more difficult.

Reply:
I'm open to other suggestions, but I failed to come up with a solution. We could use some special character based on path encoding.

Reply:
I am thinking how would this affect listing folders...

Let's assume we have this in the metainfo db:
```
<project-id>/<bucket-id>/objects/a/b/c/x_0
<project-id>/<bucket-id>/objects/a/b/c/x_1
<project-id>/<bucket-id>/objects/a/b/c/x_2
<project-id>/<bucket-id>/objects/a/b/c/y_0
<project-id>/<bucket-id>/objects/a/b/c/z_0
```
If we list with prefix `<project-id>/<bucket-id>/objects/a/b/c/`, this would return all files (x, y and z) with all their versions, right?

What if we would want to list only the latest version? There should be something that should post-process the result list, right?

What if we want to list only the latest version with limit = 2, i.e. to have x_2 and y_0 in the result? Will it be possible to do this efficiently?

Reply:
Also, just getting the latest version of an object would require listing all its versions to find the latest one, right?

We need to list with prefix like `<project-id>/<bucket-id>/objects/a/b/c/x_` and then find the greatest version. If there are thousands of versions then this would require several iterations of listing. Meaning the performance to access the latest version of an object would degrade with more versions added over time.